Installing Mods ~ Settings/Installing Mods ~ Settings.md
After running DOOMModLoader once, a "*DOOMModLoaderSettings.txt*" file will be created, which you can edit with a raw text editor to configure settings. By default, the file will look similar to this:
```cpp
{
	checkForUpdates = 30; //Check for updates before installing mods (1-30: Days between checks / 0: Always / -1: Never)
	lastUpdateCheck = "2016-05-13"; //The date of the last update check
	launchGame = true; //Launch the game after installing mods
	patchGame = true; //Patch the game to not require developer mode, if possible
	showConflicts = true; //Show mod file conflicts
	showModsInConsole = true; //Show the installed mod count in the console's footer
	showZipWarnings = false; //Show mod development warnings for zips, not just loose files
	snapMap = false; //Install mods for SnapMap instead of Campaign/Multiplayer
	uncapCutscenes = false; //Experimental: Let cutscenes run at more than 60 FPS
	verbose = false; //Display more information while installing mods
}
```
When editing this file, it's important to maintain the file's syntax. Each setting must have an equals sign between the setting's name and value, and a semicolon after the value, and all settings must be between a pair of curly brackets ( `{}` ).\
A value must be either `true`, `false`, a number, or a quoted string.\
`//` comments are supported, ignoring the rest of the line.

> [!NOTE]
> When DOOMModLoader saves its settings, any custom comments and unrecognised settings will be erased.

### `checkForUpdates`
*Default: Ask for `30` or `-1`*\
If `-1`, never checks for updates.\
If `0`, always checks for updates before installing mods.\
If `1` through `30`, this is the amount of days that must pass between successful update checks.

DOOMModLoader only checks for updates when you run it; it doesn't schedule update checks as a "service" on your system.\
No personal information will be collected. This only retrieves data from GitHub's REST API, via [**[this URL]**](https://api.github.com/repos/ZwipZwapZapony/DOOMModLoader/releases/latest).

### `lastUpdateCheck`
*Default: The date of the last update check*\
If `checkForUpdates` is `1` or higher, then `lastUpdateCheck` must be at least that many days ago for DOOMModLoader to check for updates. Otherwise, the update check will be skipped.\
If `checkForUpdates` is `-1` or `0`, then `lastUpdateCheck` will not be saved to the settings file.

### `launchGame`
*Default: Ask for `true` or `false`*\
If `true`, automatically launches DOOM (2016) or DOOM VFR after installing mods, and closes the mod loader log after a short delay if you don't pause it (on Windows).\
If `false`, doesn't launch the game.

### `patchGame`
*Default: `true`*\
If `true`, automatically patches DOOM (2016) to not require developer mode when mods are installed.\
If `false`, doesn't do that.

Without this patch, if mods are installed, then the game will crash if developer mode is disabled, or mark save slots as developer-mode-only, lock achievements/Multiplayer, and fail to save certain settings correctly (including game difficulty) if developer mode is enabled.

Before patching the game, "*-/steamapps/common/DOOM/DOOMx64(vk).exe*" will be backed up to "*-/steamapps/common/DOOM/base/DOOMx64(vk) (Pre-DOOMModLoader backup).exe*". If backups already exist, they'll be overwritten.

Setting this to `false` will not automatically unpatch the game executable; you must manually replace the game executables with the backups, or right-click DOOM (2016) in your Steam library and choose "*Properties...*" > "*Installed Files*" > "*Verify integrity of game files*".

*Special thanks to PowerBall253 ([**@brunoanc**](https://github.com/brunoanc)) for creating this patch!*

> [!NOTE]
> DOOMModLoader only patches the 2024 Steam and 2025 GOG versions of DOOM (2016) to not require developer mode. Older versions are unsupported.\
> DOOM VFR doesn't require developer mode, even without patching the game.

### `showConflicts`
*Default: `true`*\
If `true`, displays a list of conflicting files that are present in multiple mods.\
If `false`, doesn't do that.

If multiple mods contain the same file path/name, then only one version will be used, overriding all other versions. (Alphabetically-later mods will override alphabetically-earlier mods, and loose files will always override zips. E.g. "*My Super Mod.zip*" will override files in "*My Awesome Mod.zip*".)

When `showConflicts` is `true`, these files will be listed. If you're installing multiple mods and one of them doesn't work, this can make it easier to see whether that might be due to another mod.

Special files like [**[text strings]**](Creating-Mods-~-Text-Strings) or [**["*mod.decl*"]**](Creating-Mods-~-Mod.Decl) won't be listed, as they're processed individually for each mod.

### `showZipWarnings`
*Default: `false`*\
If `true`, displays all mod development warnings for all mods.\
If `false`, displays all warnings for loose mods, and less warnings for zipped mods.

This is mostly intended for mod creators.

### `snapMap`
*Default: `false`*\
If `true`, installs mods for SnapMap.\
If `false`, installs mods for Campaign/Multiplayer.

SnapMap uses a different resource container than Campaign/Multiplayer. This determines which container to install mods for, and changes the "*Installing mods...*" message to "*Installing SnapMods...*".

### `uncapCutscenes`
*Default: `false`*\
If `true`, changes cutscenes that are usually capped to 60 FPS to run at an uncapped frame rate.\
If `false`, doesn't do that.

> [!WARNING]
> This is experimental, and will likely cause problems.

### `verbose`
*Default: `false`*\
If `true`, displays more information while installing mods, such as whether a resource was added or replaced.\
If `false`, doesn't do that.

If your mod is supposed to replace a resource, but this says "*added*" instead, then you might have a typo in the file path. This is mostly intended for mod creators.

---

## Command-line arguments
All settings can also be overridden with command-line arguments, without having to edit "*DOOMModLoaderSettings.txt*" permanently. For more information, run `.\DOOMModLoader.exe -help` in a terminal/command prompt.


DOOMModLoader/Shared/Config.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Security;
using System.Text;

// Loads and saves configurable settings, plus command-line overrides for them

namespace DOOMModLoader.Shared;
static class Config
{
	public static class Cli // Command-line arguments
	{
		// Overrides for file settings
		public static bool? CheckForUpdates; // Unlike the file variant, this is only true/false
		public static bool? LaunchGame;
		public static bool? PatchGame;
		public static bool? ShowConflicts;
		public static bool? ShowZipWarnings;
		public static bool? SnapMap;
		public static bool? UncapCutscenes;
		public static bool? Verbose;

		// Command-line-exclusive options
		// "-decrypt", "-encrypt", "-extract", and "-help" are directly in the Main method
		public static bool DryRun = false; // Extract: Print would-be extracted files
		public static List<string> Filters = []; // Extract: File path/name filters
		public static bool Force = false; // Decrypt/Encrypt/Extract: Allow a non-empty destination
		public static string In = ""; // Mod directory, container to extract, or file to decrypt/encrypt
		public static byte[]? Iv = null; // Encrypt: AES IV
		public static string Out = ""; // Decrypt/Encrypt/Extract: Destination file/directory
		public static byte[]? Salt = null; // Encrypt: Salt
		public static List<string> Types = []; // Extract: Type filters
	}

	public static bool ShouldSave = false; // If false, skip saving the settings file

	public static class File // Settings that are saved to a file, and their default values
	{
		public static int? CheckForUpdates;
		public static DateOnly? LastUpdateCheck;
		public static bool? LaunchGame;
		public static bool PatchGame       = true;
		public static bool ShowConflicts   = true;
		public static bool ShowZipWarnings = false;
		public static bool SnapMap         = false;
		public static bool UncapCutscenes  = false;
		public static bool Verbose         = false;

		// Loads the settings file
		public static void Load()
		{
			ShouldSave = true;

			if (!System.IO.File.Exists("./DOOMModLoaderSettings.txt"))
			{
				// If the old settings file name exists, rename it to the new file name
				try
					{System.IO.File.Move("./DOOMModLoader Settings.txt", "./DOOMModLoaderSettings.txt");}
				catch (Exception e) when (e is DirectoryNotFoundException or FileNotFoundException
				or IOException or PathTooLongException or UnauthorizedAccessException)
					{return;} // If the old settings file name doesn't exist either, there's nothing to load
			}

			Dictionary<string, byte[]> parsed;
			try
			{
				using FileStream stream = System.IO.File.OpenRead("./DOOMModLoaderSettings.txt");
				parsed = DeclParser.Parse(stream, (int)stream.Length);
			}
			catch (FileNotFoundException)
				{return;} // The settings file doesn't exist yet, which is fine
			catch (Exception e) when (e is DirectoryNotFoundException or IOException or PathTooLongException or UnauthorizedAccessException)
			{
				Console.WriteLine();
				Prompts.WriteWarning("Warning: Failed to load \"DOOMModLoaderSettings.txt\"");
				return;
			}

			ShouldSave = false;

			byte[]? loadData; // For the LoadHelper* methods
			if (!LoadHelperI("checkForUpdates", ref File.CheckForUpdates, allowNull: true))
				ShouldSave = true;
			if (File.CheckForUpdates >= 1)
				LoadHelperD( "lastUpdateCheck", ref File.LastUpdateCheck);
			if (!LoadHelperN("launchGame",      ref File.LaunchGame, allowNull: true))
				ShouldSave = true;
			if (!LoadHelperB("patchGame",       ref File.PatchGame))
				ShouldSave = true;
			if (!LoadHelperB("showConflicts",   ref File.ShowConflicts))
				ShouldSave = true;
			if (!LoadHelperB("showZipWarnings", ref File.ShowZipWarnings))
				ShouldSave = true;
			if (!LoadHelperB("snapMap",         ref File.SnapMap))
			{
				if (System.IO.File.Exists("./base/snap_gameresources.resources")) // Don't require the SnapMap line for DOOM VFR
					ShouldSave = true;
			}
			if (!LoadHelperB("uncapCutscenes",  ref File.UncapCutscenes))
				ShouldSave = true;
			if (!LoadHelperB("verbose",         ref File.Verbose))
				ShouldSave = true;

			if (ShouldSave && parsed.ContainsKey(".error"))
			{
				// A variable wasn't found ("ShouldSave"). That's normal when updating DOOMModLoader
				// But there was also a syntax error, in which case the error is probably why it wasn't found,
				// so only show a warning when there's an error (and a missing variable)
				Console.WriteLine();
				Prompts.WriteWarning("Warning: Failed to load \"DOOMModLoaderSettings.txt\"");
			}

			// String-to-value shortcuts. Sets a value and returns true if the value was found, false otherwise
			bool LoadHelperB(string name, ref bool result, bool allowNull = false)
			{
				if (!parsed.TryGetValue(name, out loadData))
					return false;
				else if (MemoryExtensions.SequenceEqual(loadData, "true"u8))
				{
					result = true;
					return true;
				}
				else if (MemoryExtensions.SequenceEqual(loadData, "false"u8))
				{
					result = false;
					return true;
				}
				return (allowNull && MemoryExtensions.SequenceEqual(loadData, "NULL"u8));
			}
			bool LoadHelperN(string name, ref bool? result, bool allowNull = false)
			{
				if (!parsed.TryGetValue(name, out loadData))
					return false;
				else if (MemoryExtensions.SequenceEqual(loadData, "true"u8))
				{
					result = true;
					return true;
				}
				else if (MemoryExtensions.SequenceEqual(loadData, "false"u8))
				{
					result = false;
					return true;
				}
				return (allowNull && MemoryExtensions.SequenceEqual(loadData, "NULL"u8));
			}
			bool LoadHelperD(string name, ref DateOnly? result, bool allowNull = false)
			{
				if (!parsed.TryGetValue(name, out loadData))
					return false;
				string text = Encoding.UTF8.GetString(loadData);
#pragma warning disable IDE0018 // "Variable declaration can be inlined"
				DateOnly temp;
#pragma warning restore IDE0018
				if (DateOnly.TryParseExact(text, "\\\"yyyy-MM-dd\\\"", CultureInfo.InvariantCulture, DateTimeStyles.None, out temp))
				{
					result = temp;
					return true;
				}
				return (allowNull && text == "NULL");
			}
			bool LoadHelperI(string name, ref int? result, bool allowNull = false)
			{
				if (!parsed.TryGetValue(name, out loadData))
					return false;
#pragma warning disable IDE0018 // "Variable declaration can be inlined"
				int temp;
#pragma warning restore IDE0018
				if (int.TryParse(loadData, out temp))
				{
					if (temp is >= -1 and <= 30) // Only allow up to a month between update checks
					{
						result = temp;
						return true;
					}
					else
						return false;
				}
				return (allowNull && MemoryExtensions.SequenceEqual(loadData, "NULL"u8));
			}
		}

		// Writes the settings file
		public static void Save()
		{
			if (!ShouldSave)
				return;
			ShouldSave = false; // Set it to false even if the file fails to save

			string text = "{"
			+ $"\n\tcheckForUpdates = {ValueI(File.CheckForUpdates)};"
			+  " //Check for updates before installing mods (1-30: Days between checks / 0: Always / -1: Never)";
			if (File.CheckForUpdates >= 1 && (File.LastUpdateCheck is not null))
			{
				text += $"\n\tlastUpdateCheck = {ValueD(File.LastUpdateCheck)};"
				+  " //The date of the last update check";
			}
			text += $"\n\tlaunchGame = {ValueB(File.LaunchGame)};"
			+  " //Launch the game after installing mods"
			+ $"\n\tpatchGame = {ValueB(File.PatchGame)};"
			+  " //Patch the game to not require developer mode, if possible"
			+ $"\n\tshowConflicts = {ValueB(File.ShowConflicts)};"
			+  " //Show mod file conflicts"
			+ $"\n\tshowZipWarnings = {ValueB(File.ShowZipWarnings)};"
			+  " //Show mod development warnings for zips, not just loose files";
			if (System.IO.File.Exists("./base/snap_gameresources.resources")) // Don't write the SnapMap line for DOOM VFR
			{
				text += $"\n\tsnapMap = {ValueB(File.SnapMap)};"
				+  " //Install mods for SnapMap instead of Campaign/Multiplayer";
			}
			text += $"\n\tuncapCutscenes = {ValueB(File.UncapCutscenes)};"
			+  " //Experimental: Let cutscenes run at more than 60 FPS"
			+ $"\n\tverbose = {ValueB(File.Verbose)};"
			+  " //Display more information while installing mods"
			+  "\n}";

			try
				{System.IO.File.WriteAllText("./DOOMModLoaderSettings.txt", text);}
			catch (Exception e) when (e is DirectoryNotFoundException or IOException or NotSupportedException
			or PathTooLongException or SecurityException or UnauthorizedAccessException)
				{Prompts.WriteWarning("Warning: Failed to save \"DOOMModLoaderSettings.txt\"");}

			// Value-to-string shortcuts
			static string ValueB(bool? x) => (x is null) ? "NULL" : (x == true ? "true" : "false");
			static string ValueD(DateOnly? x) => (x is null) ? "NULL" : $"\"{x.Value.Year}-{x.Value.Month:D2}-{x.Value.Day:D2}\"";
			static string ValueI(int? x) => (x is null) ? "NULL" : x.Value.ToString();
		}
	}

	public static class Final // Command-line arguments, falling back to file settings if not overridden
	{
		public static int? CheckForUpdates
		{
			get
			{
				if (Cli.CheckForUpdates is null)
					return File.CheckForUpdates;
				else if (Cli.CheckForUpdates == true)
					return 0; // Always
				else
					return -1; // Never
			}
		}
		public static bool? LaunchGame     => (Cli.LaunchGame      ?? File.LaunchGame);
		public static bool PatchGame       => (Cli.PatchGame       ?? File.PatchGame);
		public static bool ShowConflicts   => (Cli.ShowConflicts   ?? File.ShowConflicts);
		public static bool ShowZipWarnings => (Cli.ShowZipWarnings ?? File.ShowZipWarnings);
		public static bool SnapMap         => (Cli.SnapMap         ?? File.SnapMap);
		public static bool UncapCutscenes  => (Cli.UncapCutscenes  ?? File.UncapCutscenes);
		public static bool Verbose         => (Cli.Verbose         ?? File.Verbose);
	}
}
